### PR TITLE
`cmp_client.c`: fix handling of total_timeout for RR and GENM transactions

### PR DESCRIPTION
**Pull request by DDvO**
It turns out that initialization of `ctx->end_time` was missing for revocation and general message transactions, which mislead the timeout checks and error reporting.

Also fix related documentation.

### CODE REVIEW

**Comment by t8m** [comment]
@DDvO would it be possible to add a test for this?

**Comment by DDvO** [comment]
> @DDvO would it be possible to add a test for this?

Good question.
So far, it is not directly possible. Yet after thinking more about it, I found a way of doing so.

The point is that one needs to be able to provoke a timeout condition also for RR and GENM transactions,
which (in contrast to IR/CR/KUR, i.e., cert enrollment transactions) in CMPv2 cannot involve polling,
such that in tests it is not possible to add client-side delays. Also the test server does not provide delays.

My solution is to make the internal handling of the CMP context `total_timeout` field more flexible,
allowing to inject in tests negative values, which can then be used to provoke an immediate timeout condition.
Doing so,
* I made sure not to change the API and CLI behavior
* I further improved the documentation by clarifying the behavior on (negative) message/total timeout input values.
* after extending the tests, I found that the fix so far provided here was insufficient, and extended the fix.

The extended tests build on #19760 because otherwise ugly merge conflicts would occur.
Since #19760 has already been approved, I suggest that I wait with pushing the extensions of this PR
until it can be rebased on the merged #19760 tomorrow.

**Comment by DDvO** [comment]
Or can I rebase this PR on #19760 already now?

**Comment by t8m** [comment]
You can just wait. No hurry.

**Comment by DDvO** [comment]
@openssl/committers can we please get this done?
2nd approval is still pending.

**Review by hlandau — APPROVED** [review]

**Comment by DDvO** [comment]
Merged to master, 3.0, and 3.1
Thanks @t8m and @hlandau

**Pull request closed without merging**
